cache/local_session.cpp

```diff
@@ -248,30 +248,25 @@ int local_session::write(const dnet_id &id,
 	return err;
 }
 
-std::unique_ptr<ioremap::elliptics::n2::lookup_response> local_session::lookup(const dnet_cmd &tmp_cmd, int *errp)
+std::shared_ptr<n2::lookup_response> local_session::lookup(const dnet_cmd &tmp_cmd, int *errp)
 {
-	std::unique_ptr<n2::lookup_request> request(new(std::nothrow) n2::lookup_request(tmp_cmd));
-	if (!request) {
-		*errp = -ENOMEM;
-		return nullptr;
-	}
-	request->cmd.flags |= m_cflags;
-	request->cmd.size = 0;
-	request->cmd.backend_id = m_backend.backend_id();
+	n2_request request(tmp_cmd, n2::default_deadline());
+	request.cmd.flags |= m_cflags;
+	request.cmd.size = 0;
+	request.cmd.backend_id = m_backend.backend_id();
 
-	std::unique_ptr<n2::lookup_response> response;
+	std::shared_ptr<n2::lookup_response> response;
 	n2_repliers repliers{
-		[&](std::unique_ptr<n2_message> message) {
-			response.reset(static_cast<n2::lookup_response *>(message.release()));
+		[&](const std::shared_ptr<void> &resp) {
```

Comment:
`std::shared_ptr<void>` looks weird and unsafe. Why was `message` removed?

Reply:
We made downcast from `message*` via `static_cast` before. And now we will make downcast from `void*` via `static_cast`. I don't see difference :)

Reply:
> Why was message removed?

We could use `class n2_body { virtual ~n2_body() = default; }` as base. There are no common members between message bodies. `std::shared_ptr<void>` now describes only body.

Reply:
Fixed https://github.com/interiorem/elliptics/pull/272/commits/25a3a37a4ec3e6b3e899b7bae63842e7ff7d6912